text/0000-rustc-bootstrap.md

```diff
@@ -0,0 +1,164 @@
+- Feature Name: `RUSTC_ALLOW_UNSTABLE`
+- Start Date: 2025-05-18
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes decoupling the two components of our stability policy: still requiring feature gates, but allowing feature gates to be enabled on stable. It does so by adding `RUSTC_ALLOW_UNSTABLE_<feature_name>` environment variables which can be used to permit using feature gates on stable toolchain channels. These variable are respected by all official Rust tools that use feature gates.
+
+This RFC is *not* intended as a general purpose mechanism for Rust developers to use nightly features on stable; it's specifically targeted at build systems wrapping cargo, such as distro packagers, external tools shipped with the toolchain, and large projects that build a custom Rust toolchain from source.
+
+# Motivation
+[motivation]: #motivation
+
+## Why allow using unstable features on stable?
+
+Rust's stability policy has two components:
+1. To the extent possible, each unstable feature comes with a feature gate, and is disabled when that feature gate is inactive. [^1]
+2. Enabling feature gates is only allowed on the nightly toolchain.
+
+[^1]: There are some exceptions to this, such as https://github.com/rust-lang/rust/issues/139892#issuecomment-2808505610. But in general we attempt to make sure all unstable features have a feature gate.
+
+Our motivation for 1 (having feature gates) is to make sure that people do not unknowingly rely on unstable features. This was a big problem for e.g. intra-doc links, which [people often used without knowing they were unstable][63305], making it impossible to remove the feature.
+
+[63305]: https://github.com/rust-lang/rust/issues/63305
+
+Our motivation for 2 (disabling feature gates on stable) is three-fold:
+1. Prevent people from relying on features that may change in the future while on the stable toolchain, upholding our "stability without stagnation" motto.
+2. Disallow library authors from "silently" opt-ing in to unstable features, such that the person running the top-level build doesn't know they're using unstable features that may break when the toolchain is updated. This rationale doesn't apply to nightly, where the party running the top-level build is assumed to know that nightly comes with no stability guarantees.
```

Comment:
I think Cargo already blocks build scripts from exporting RUSTC_BOOTSTRAP.

Reply:
If we allowed feature gates on stable without what is proposed in this RFC, then Cargo can't block dependencies from using unstable features as `RUSTC_BOOTSTRAP` wouldn't be necessary to set.

Reply:
There are two different things here. First, the current behavior. @celinval is correct that Cargo blocks build scripts from setting RUSTC_BOOTSTRAP ([I wrote that PR](https://github.com/rust-lang/cargo/pull/9181)).

Secondly, the new behavior. I think we should continue to disallow build scripts to silently opt-in, which means that Cargo should check if the env variable it's trying to set starts with `RUSTC_ALLOW_UNSTABLE_`, and if so, disallow it. That's equivalent to the current situation today, there's no change there.